File: src/chain-observers/collectors/jsonrpc/decline-voting-rights-collector.ts
import { DeclineVotingRightsClassifier } from "../../classifiers";
import { IDeclinedVotingRightsAccount } from "../../classifiers/decline-voting-rights-classifier";
import { DataEvaluationContext } from "../../factories/data-evaluation-context";
import { CollectorBase, TAvailableClassifiers } from "../collector-base";


const MAX_DECLINED_VOTING_RIGHTS_GET_LIMIT = 1000;

export class DeclineVotingRightsCollector extends CollectorBase<DeclineVotingRightsClassifier> {
  private readonly declineVotingRightsAccounts: Record<string, number> = {};

  protected pushOptions(data: DeclineVotingRightsClassifier["optionsType"]): void {
    this.declineVotingRightsAccounts[data.declineVotingRightsAccount] = (this.declineVotingRightsAccounts[data.declineVotingRightsAccount] || 0) + 1;
  }

  protected popOptions(data: DeclineVotingRightsClassifier["optionsType"]): void {
    this.declineVotingRightsAccounts[data.declineVotingRightsAccount] = (this.declineVotingRightsAccounts[data.declineVotingRightsAccount] || 1) - 1;

    if (this.declineVotingRightsAccounts[data.declineVotingRightsAccount] === 0)
      delete this.declineVotingRightsAccounts[data.declineVotingRightsAccount];
  }

  public async get(data: DataEvaluationContext) {
    const declineVotingRightsAccounts: Record<string, IDeclinedVotingRightsAccount> = {};

    const recoveryAccounts = Object.keys(this.declineVotingRightsAccounts);
    for (let i = 0; i < recoveryAccounts.length; i += MAX_DECLINED_VOTING_RIGHTS_GET_LIMIT) {
      const chunk = recoveryAccounts.slice(i, i + MAX_DECLINED_VOTING_RIGHTS_GET_LIMIT);

      const startFindDeclineVotingRightsRequests = Date.now();
      const { requests } = await this.worker.chain!.api.database_api.find_decline_voting_rights_requests({ accounts: chunk });
      data.addTiming("database_api.find_decline_voting_rights_requests", Date.now() - startFindDeclineVotingRightsRequests);

      for(const request of requests)
        declineVotingRightsAccounts[request.account] = {
          account: request.account,
          effectiveDate: new Date(`${request.effective_date}Z`)
        };
    }

    return {
      /*
       * Instruct TypeScript typings that DeclineVotingRightsClassifier.name is actualy a Classifier name we expect.
       * This is required for the bundlers to properly deduce the type of the classifier in data evaluation context.
       */
      [DeclineVotingRightsClassifier.name as "DeclineVotingRightsClassifier"]: {
        declineVotingRightsAccounts
      } as TAvailableClassifiers["DeclineVotingRightsClassifier"]
    };
  };
}

File: python/workerbee/chain_observers/collectors/jsonrpc/decline_voting_rights_collector.py
"""DeclineVotingRightsCollector — fetches decline voting rights requests."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ...classifiers.decline_voting_rights_classifier import DeclineVotingRightsClassifier
from ..set_managed_collector import SetManagedCollector

if TYPE_CHECKING:
    from ...factories.data_evaluation_context import TCollectorEvaluationContext


class DeclineVotingRightsCollector(SetManagedCollector):
    _options_key = "decline_voting_rights_account"

    async def get(self, data: TCollectorEvaluationContext) -> dict[str, Any]:
        if not self._tracked:
            return {DeclineVotingRightsClassifier.__name__: {"decline_voting_rights_accounts": {}}}

        accounts: dict[str, Any] = {}
        for chunk in self._tracked_chunks():
            with data.add_timing("database_api.find_decline_voting_rights_requests"):
                result = await self.worker.chain.api.database_api.find_decline_voting_rights_requests(
                    accounts=chunk,
                )

            for req in result.requests:
                accounts[req.account] = {
                    "account": req.account,
                    "effective_date": req.effective_date,
                }

        return {DeclineVotingRightsClassifier.__name__: {"decline_voting_rights_accounts": accounts}}
